iml-services/iml-stats/src/main.rs

```diff
@@ -531,51 +532,97 @@ async fn main() -> Result<(), ImlStatsError> {
         .query_into(
             format!(
                 r#"
-            SELECT mgs_fs,is_mgs_fs
-            FROM target
-            WHERE mgs_fs='{}' AND is_mgs_fs=true
-            ORDER BY time ASC
-            LIMIT 2"#,
-                fs_names
+                SELECT mgs_fs, mgs_fs_count FROM
+                    (SELECT count(is_mgs_fs) AS mgs_fs_count FROM target WHERE is_mgs_fs=true GROUP BY mgs_fs),
+                    (SELECT LAST(is_mgs_fs),time as last_time FROM target GROUP BY mgs_fs)
+                GROUP BY mgs_fs
+                "#
             )
             .as_str(),
-            Some(Precision::Nanoseconds),
+            Some(Precision::Nanoseconds)
         )
         .await?
         .unwrap_or_default();
 
-    if xs.len() < 2 {
-        return Ok(());
-    }
+    tracing::debug!(
+        "delete stale mgs_fs records - fs_names: {:?}; query results: {:?}",
+        fs_names,
+        xs
+    );
 
-    if let Some(x) = xs.get(0) {
-        let query = format!(
-            "DELETE FROM target WHERE time < {} AND mgs_fs = '{}' AND kind = '{}'",
-            x.time,
-            fs_names,
-            TargetVariant::MGT
-        );
+    let subsets: HashMap<String, (u64, u32)> = xs
+        .into_iter()
+        .filter(|x| {
+            let mgs_fs_set = fs_names_to_hash_set(&x.mgs_fs);
 
-        tracing::debug!("Running query: {}", query);
+            let intersection = mgs_fs_set.intersection(&fs_names).collect::<Vec<_>>();
 
-        client
-            .query(query.as_str(), Some(Precision::Nanoseconds))
-            .map_err(|e| iml_influx::Error::InfluxDbError(e))
-            .await?;
-    }
+            !intersection.is_empty()
+        })
+        .fold(HashMap::new(), |mut acc, record| {
+            let mgs_fs = record.mgs_fs;
+            let mut entry = acc.entry(mgs_fs).or_insert((0, 0));
+
+            if record.time > 0 {
+                entry.0 = record.time;
+            }
+
+            if let Some(count) = record.mgs_fs_count {
+                entry.1 = count;
+            }
+
+            acc
+        });
+
+    let delete_subset_futures = subsets
+        .into_iter()
+        .filter(|(mgs_fs, (_time, count))| {
+            let mgs_fs_set = fs_names_to_hash_set(&mgs_fs);
+
+            mgs_fs_set != fs_names || *count > 1
+        })
+        .map(|(mgs_fs, (time, count))| {
+            let mds_fs_set = fs_names_to_hash_set(&mgs_fs);
+
+            let query = if mds_fs_set == fs_names && count > 1 {
+                format!(
+                    "DELETE FROM target WHERE time < {} AND mgs_fs = '{}' AND kind = '{}'",
+                    time,
+                    mgs_fs,
+                    TargetVariant::MGT
+                )
+            } else {
+                format!(
+                    "DELETE FROM target WHERE time <= {} AND mgs_fs = '{}' AND kind = '{}'",
+                    time,
+                    mgs_fs,
+                    TargetVariant::MGT
```

Comment:
The queries look the same except `time <= {}`.

Reply:
Yes, in the case where the sets are equal we want to delete all entries except for the latest entry. In the case where we are dealing with a subset, we want to delete all entries <= the latest time.